src/apiv2.ts

```diff
@@ -0,0 +1,283 @@
+    }
+    try {
+      return this.doRequest<ReqT, ResT>(reqOptions);
+    } catch (err) {
+      if (err instanceof FirebaseError) {
+        throw err;
+      }
+      throw new FirebaseError(`Failed to make request: ${err}`, { original: err });
+    }
+  }
+
+  private addRequestHeaders<T>(reqOptions: ClientRequestOptions<T>): ClientRequestOptions<T> {
+    if (!reqOptions.headers) {
+      reqOptions.headers = {};
+    }
+    reqOptions.headers["Connection"] = "keep-alive";
+    reqOptions.headers["User-Agent"] = `FirebaseCLI/${CLI_VERSION}`;
+    reqOptions.headers["X-Client-Version"] = `FirebaseCLI/${CLI_VERSION}`;
+    if (reqOptions.responseType === "json") {
+      reqOptions.headers["Content-Type"] = "application/json";
+    }
+    return reqOptions;
+  }
+
+  private async addAuthHeader<T>(
+    reqOptions: ClientRequestOptions<T>
+  ): Promise<ClientRequestOptions<T>> {
+    if (!reqOptions.headers) {
+      reqOptions.headers = {};
+    }
+    const token = await this.getAccessToken();
+    reqOptions.headers["Authorization"] = `Bearer ${token}`;
+    return reqOptions;
+  }
+
+  private async getAccessToken(): Promise<string> {
+    // Runtime fetch of Auth singleton to prevent circular module dependencies
+    if (accessToken) {
+      return accessToken;
+    }
+    const data = await require("./auth").getAccessToken(refreshToken, []);
+    return data.access_token;
+  }
+
+  private async doRequest<ReqT, ResT>(
+    options: ClientRequestOptions<ReqT>
+  ): Promise<ClientResponse<ResT>> {
+    if (!options.path.startsWith("/")) {
+      options.path = "/" + options.path;
+    }
+
+    const versionPath = this.opts.apiVersion ? `/${this.opts.apiVersion}` : "";
+    let fetchURL = `${this.opts.urlPrefix}${versionPath}${options.path}`;
+    if (options.queryParams) {
+      // TODO(bkendall): replace this half-hearted implementation with
+      // URLSearchParams when on node >= 10.
+      const sp: string[] = [];
+      for (const key of Object.keys(options.queryParams)) {
+        const value = options.queryParams[key];
+        sp.push(`${key}=${encodeURIComponent(value)}`);
+      }
+      if (sp.length) {
+        fetchURL += "?" + sp.join("&");
+      }
+    }
+
+    const fetchOptions: RequestInit = {
+      headers: options.headers,
+      method: options.method,
+    };
+
+    if (options.json) {
+      fetchOptions.body = JSON.stringify(options.json);
+    }
+
+    this.logRequest(options);
+
+    let res: Response;
+    try {
+      res = await fetch(fetchURL, fetchOptions);
+    } catch (err) {
+      throw new FirebaseError(`Failed to make request to ${fetchURL}`, { original: err });
+    }
+
+    let body: ResT;
+    if (options.responseType === "json") {
+      body = await res.json();
+    } else {
+      // This is how the linter wants the casting to T to work.
+      body = ((await res.text()) as unknown) as ResT;
+    }
+
+    this.logResponse(res, body, options);
+
+    if (res.status >= 400) {
```

Comment:
This shouldn't throw if `resolveOnHttpError` is in the options.

Reply:
Done. Was going to say "nothing uses this yet, I'll add a todo", but it's so simple it's not worth that effort haha